proj/src/main/webapp/script.js

```diff
@@ -27,17 +29,101 @@ if(btnLogin){
     });
 }
 
+
+// Adds user to an existing chat when given a reference to the place in the database
+function addUserToTag(reference){
+    var currentUID = firebase.auth().currentUser.uid
+    console.log("adding new user to chat room with uid: " + currentUID);
+    reference.push(currentUID);
+}
+
+// Creates a new chat given a tag and adds the current user as a member
+function createNewChatWithUser(tag){
+    console.log("creating new chat with tag: " + tag);
+    var newChat = {
+        "name" : tag,
+        "tag" : tag,
+    };
+    var currentReference = firebase.database().ref("/chat/" + tag);
+    var postKey = currentReference.push(newChat).key;
+    currentReference = firebase.database().ref("/chat/" + tag + "/" + postKey + "/users/");
+    addUserToTag(currentReference);
+
+    return postKey;
+}
+
+// Loops through open chat rooms, adds the user to the first open chat and returns the key
+function findChatAndAddUser(snapshot){
+    var key;
+    var foundOpenRoom = false;
+    snapshot.forEach(function(childSnapshot){
+        key = childSnapshot.key;
+        console.log("key1 " + key);
+
+        var userSnapshot = childSnapshot.child("users");
+        var usersReference = userSnapshot.ref;
+
+        // Handled below if spillover is needed
+        if(userSnapshot.numChildren() < MAX_CHAT_SIZE){
+            addUserToTag(usersReference);
+            foundOpenRoom = true;
+            return true;
+        }
+    });
+    if(foundOpenRoom){
+        return key;
+    } else {
+        // Create new chat if room is full, add new user
+        return createNewChatWithUser(currentTag);
+    }
+}
+
+// Create or join chatroom
+function createOrJoinChat(currentTag){
+    var ref = firebase.database().ref("/chat/");
+    return ref.once("value").then(function(snapshot){
+        // Checks to see if tag already exists in database
+        if(snapshot.hasChild(currentTag)){
+
+            // If tag already exists, navigate to users and add users if there is room
+            var query = firebase.database().ref("/chat/" + currentTag + "/").orderByKey();
```

Comment:
Maybe the same comment here depending on what the helper method was. Or maybe you can make a similar helper method for getting the reference to the tag and then the method for getting the reference to the key can build off of it.

Reply:
Okay so I get what you're saying, but this seems pretty minor, and would not save terribly much. Either way, I'm providing a new path and would need to change it in all places that I'm getting a new reference; either as part of concatenation or as parameters to a function call.

Reply:
Arinze, I know you like having a lot of helper methods in your code, but I think this is all right.